r1cs/gadgets/std/src/fields/nonnative/nonnative_field_mul_result_gadget.rs

```diff
@@ -205,13 +324,27 @@ impl<SimulationF: PrimeField, ConstraintF: PrimeField>
             }
         }
 
+        // Compute the limbs of `k * p + r`. Does not cost any constraint.
+        // TODO: Let us remove the `num_adds` declaration, as it is not used anyway.
+        // The `k*p + r` limbs are bounded by 
+        // ``
+        //      (k * p + r)[i] <= (num_limbs + num_add)* 2^{2*bits_per_limb[i]} 
+        //                      + 2^bits_per_limb[i] 
+        //                      <= (num_limbs + num_add + 1) * 2^{2*bits_per_limb[i]}. 
+        // ``
+        // Hence we may set num_adds and surfeit for the limbs of `k*p + r` according
+        // to
+        // ``
+        //      num_adds(kp + r) = num_limbs + num_add,
+        //      surfeit(kp + r) = log(num_limbs + num_adds + 1).
+        // ``
```

Comment:
This can be explained more precisely in my opinion:
the addends limb of   k * p   are at (num_limbs - 1) which are product of two limbs in normal form
 and at most one which is a product of one in normal form and one oversized.
(k * p + r)[i] <= 2^bits_per_limb -1   +   (num_limbs - 1) * (2^bits_per_limb - 1)^2 
                + (2^bits_per_limb - 1) * (2^(bits_per_limb + len(num_adds + 1) + 1) - 1)
                     <  (num_limbs - 1) * 2^(2 * bits_per_limb ) + 2^(2 * bits_per_limb + len(num_adds + 1) + 1)
                     =  2^(2 * bits_per_limb) * ( num_limbs - 1 + 2^( len(num_adds + 1) ) + 1)
From this it would follow that surfeit(k*p + r) = len(num_limbs + 2^( len(num_adds + 1) ).
WARNING: Note that this can be bigger than len(num_limbs + num_adds + 1).

Reply:
As discussed offline, the limbs of kp are strictly bounded by `num_adds + 1 + num_limbs -1` times `2^{2*bits_per_limb}`. The `+1` in the stated formula stems from the addition with `r`.